SoundWire clock stop support

plbossart opened the pull request:
This is a proposal to allow the clock stop behavior to be modifiable as needed, depending on the DSP driver requirements

@RanderWang can you try and add support for the clock stop mode with the bus reset on restart? We can enable the other modes when we have D0i3 support, or figure out a means to keep the parent active when the clock stops.

Comments welcome.

bardliao commented:
@plbossart Can we use different clock_stop_quirk for each link?

plbossart commented:
> @plbossart Can we use different clock_stop_quirk for each link?

I thought about it but I am not sure if it's useful. The power dependencies are really between the DSP and the SoundWire IPs, if you want to enable fast restart on one link then the DSP needs to remain active (D0/D0i3), so you might as well enable fast restart on all links. Conversely if you do want to turn the DSP off for power consumption reasons, then all links needs to restart with a bus reset.

The only thing that might be useful is e.g. if there are latency requirements for specific links, but the platform driver does not know this, it's handled at the machine driver level which is the only entity that might know about such requirements.

we can change this later if needed, the code in intel.c does not presume anything about other links, it'd just be the interface between SOF/hda and intel_init that would change. I think it's already be very good if we have several knobs to play with.

RanderWang commented:
> This is a proposal to allow the clock stop behavior to be modifiable as needed, depending on the DSP driver requirements
> 
> @RanderWang can you try and add support for the clock stop mode with the bus reset on restart? We can enable the other modes when we have D0i3 support, or figure out a means to keep the parent active when the clock stops.

 I also did some tests.  I tried to add a flag PCI_DEV_FLAGS_NO_D3 to make audio pci device in D0 when it is inactive, but master was still powered off. Later I found master will be powered off if dsp core is reset and powered down. I don't why, maybe it is a hw design.

I will finish the clock stop mode tomorrow for I am blocked by other issues.

plbossart commented:
> > This is a proposal to allow the clock stop behavior to be modifiable as needed, depending on the DSP driver requirements
> > @RanderWang can you try and add support for the clock stop mode with the bus reset on restart? We can enable the other modes when we have D0i3 support, or figure out a means to keep the parent active when the clock stops.
> 
> I also did some tests. I tried to add a flag PCI_DEV_FLAGS_NO_D3 to make audio pci device in D0 when it is inactive, but master was still powered off. Later I found master will be powered off if dsp core is reset and powered down. I don't why, maybe it is a hw design.
> 
> I will finish the clock stop mode tomorrow for I am blocked by other issues.

Thanks @RanderWang. I'd like to provide an update for the SoundWire interfaces later this week or early next week, so if you can check the definitions that would be nice.

RanderWang commented:
@plbossart   I updated a prototype https://github.com/plbossart/sound/pull/7. Thanks!

plbossart commented:
@RanderWang @bardliao @ranj063 @YvonneYang2 can you take a look at this updated PR? Basically it's a merge/cleanup of Rander's work with my initial definitions.

I tested the NOT_ALLOWED mode, which keeps the master active.
I tested the TEARDOWN mode, which essentially prevents any jack detection (same as before)
I was not successful on Olympic with the BUS_RESET mode. Somehow the master-2 fails, I am not sure why the configuration to accept CMD_IGNORED doesn't seem to work.

````
[   24.007288] intel-sdw sdw-master-2: intel_suspend_runtime start
[   24.007297] intel-sdw sdw-master-2: sdw_cdns_clock_stop: start
[   24.007303] intel-sdw sdw-master-2: sdw_cdns_clock_stop: slave attached 0
[   24.007498] intel-sdw sdw-master-2: Msg ignored for Slave 15
[   24.007501] intel-sdw sdw-master-2: ClockStopNow Broadcast message failed -61
[   24.007505] intel-sdw sdw-master-2: bus clock stop failed -61
[   24.007508] intel-sdw sdw-master-2: cannot enable clock stop on suspend
````

@RanderWang did you test on RVPs or on Olympic devices? @bardliao this probably works on other devices with 2 amps or the 3-in-1 board.

Edit: if I mask the link 2 with 
options soundwire_intel_init sdw_link_mask=0xB
and change the acpi matching table, I am able to test jack detection while we are in D3/pm_runtime_suspend. We just need to figure out how to stop the clock on a link that have no slaves (that may be already solved with newer BIOS but it's still a problem to deal with).

Also I am not sure I understand why we added a 'block wake' parameter, the documentation states that the block_wake needs to be set prior to entering clock_stop mode?

Last, I added a tentative support for the regular mode where we keep the PCI  device active. I didn't test since I am not sure the sequences are fine.

I think we are very close to the goal here, hopefully we've got enough configurations to please everyone, and we'll have even better results when D0i3 is supported

To test, you can use 

options snd-sof-intel-hda-common sdw_clock_stop_quirks=0x2  for teardown
options snd-sof-intel-hda-common sdw_clock_stop_quirks=0x1  for not allowed
options snd-sof-intel-hda-common sdw_clock_stop_quirks=0x8  for bus reset
options snd-sof-intel-hda-common sdw_clock_stop_quirks=0x0  for normal mode

please do not play with the sof-debug options or disable pm_runtime by hand with module parameters.

RanderWang commented on drivers/soundwire/intel.c:
```diff
@@ -1268,15 +1268,12 @@ static struct sdw_master_ops sdw_intel_ops = {
 
-	if (!init_master)
-		return 0;
-
```
@plbossart  it is not always true.  If audio pci device is in D0 or D0i3,  master can preserve its context, so we don't need to reinitialize it again.  If audio pci device is in D0 or D0i3 and master is in clock stop state, we can't update config before clearing clock_stop in MCP_CONTROL, so sdw_cdns_init will return config update timeout.

RanderWang replied:
hi @plbossart ,  I make another PR to fix a io issue found by QA on this branch. please 
 help to review it it https://github.com/plbossart/sound/pull/8. Thanks!  
We don't need to revert above change if this PR is used.

plbossart replied:
@RanderWang I merged all your code from plbossart#8 in this PR (with a couple of cleanups)

YvonneYang2 commented:
With options snd-sof-intel-hda-common sdw_clock_stop_quirks=0x8 for bus reset, switch output source to headphone in sound setting during paplay will print dmesg like:
```
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229348] intel-sdw sdw-master-0: intel_prepare: SDW0 Pin2: start
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229353] rt711 sdw:0:25d:711:0: DPN_PortCtrl register write failed for port 3
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229355] intel-sdw sdw-master-0: Program transport params failed: -22
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229357] intel-sdw sdw-master-0: Program params failed: -22
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229359] sdw_prepare_stream: SDW0 Pin2-Playback: done
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229360] intel-sdw sdw-master-0: intel_prepare: SDW0 Pin2: done
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229362] intel-sdw sdw-master-0: ASoC: cpu DAI prepare error: -22
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229364]  SDW0-Playback: ASoC: backend prepare failed -22
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229390]  Headphone: ASoC: hw_free FE Headphone
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229392] sof-audio-pci 0000:00:1f.3: pcm: free stream 0 dir 0
Oct 16 06:54:32 ubuntu-Latitude-11111 kernel: [   52.229395] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
```
Tested it with https://github.com/plbossart/sound/pull/8, issue cannot be reproduced. 

Logs: 
[source_switch.zip](https://github.com/thesofproject/linux/files/3931284/source_switch.zip)

Note: 
After suspend/resume stress test 300 times, then aplay0,2(RT1308) will print logs seems same as above. Will test it again with https://github.com/plbossart/sound/pull/8 to check if it can fix the issue.

Log of aplay0,2 failed after suspend/resume stress test:
[dmesg.log](https://github.com/thesofproject/linux/files/3931693/dmesg.log)

RanderWang commented:
@plbossart   if clock_stop_quirks == 0, our driver doesn't work. According to spec, we need glue hw involved to enable clock stop. First I tuned the driver without glue hw for a long time, but I still get io timeout. And with glue hw involved,  there is no problem.  And for this case, pci driver is in D0, we need to process wakeen interrupt in sof pci thread.

I update all my change in https://github.com/plbossart/sound/pull/8.  please check and merge some patch into one. Thanks!

ranj063 commented on drivers/soundwire/cadence_master.c:
```diff
@@ -224,12 +224,30 @@ static int cdns_clear_bit(struct sdw_cdns *cdns, int offset, u32 value)
 
 		timeout--;
-		udelay(50);
+		usleep_range(50, 100);
```
The patch is fine but the commit message has typos: "Parameters are provided to prevent
Slave-initiated wakes, which make be required in specific cases, and
well as controlling the restart mode." -> "Parameters are provided to prevent
Slave-initiated wakes which may be required in specific cases as well as controlling the restart mode."

Also, it is not quite clear from the commit message what "parameters" you're talking about

plbossart replied:
reworded commit message

ranj063 commented on drivers/soundwire/intel.c:
```diff
@@ -1644,9 +1653,9 @@ static int intel_resume_runtime(struct device *dev)
+		ret = sdw_cdns_clock_restart(cdns, true);
 		if (ret < 0) {
-			dev_err(dev, "unable to exit bus reset sequence during resume\n");
+			dev_err(dev, "unable to resume master during resume\n");
```
unable to resume master during resume? maybe just "unable to resume master"?

ranj063 commented on drivers/soundwire/intel.c:
```diff
@@ -446,6 +446,44 @@ static int intel_link_power_down(struct sdw_intel *sdw)
+		intel_shim_wake(sdw, false);
+
+		/*
+		 * wake up master and slave so that slave can notify master
```
we're only waking up the slave here right? So the comment "wake up master ..." is not correct as you dont wake master up explicitly here?

plbossart replied:
@RanderWang this piece of code is indeed troubling.

I am not sure it's quite right to test for the slave status when handling the wakeen events, since the IP may be lost and the slave status is meaningless until the bus has actually restarted.

I think what you were trying to do here is to only resume actual devices that can generate a wake and ignore both the devices that cannot generate a wake and the ghost devices present in ACPI tables only, but the use of the status is very dangerous and racy.

RanderWang replied:
> we're only waking up the slave here right? So the comment "wake up master ..." is not correct as you dont wake master up explicitly here?

yes, not explicitly, just implicitly.  If slave is waken up, the parent master should be waken up.

RanderWang replied:
> @RanderWang this piece of code is indeed troubling.
> 
> I am not sure it's quite right to test for the slave status when handling the wakeen events, since the IP may be lost and the slave status is meaningless until the bus has actually restarted.

If the slave is unattached before clock stop, in my opinion, It can't wake up master with clock stop. 
> 
> I think what you were trying to do here is to only resume actual devices that can generate a wake and ignore both the devices that cannot generate a wake and the ghost devices present in ACPI tables only, but the use of the status is very dangerous and racy.

yes, you got my idea. I don't want to wake up any ghost device.

RanderWang replied:
@plbossart  I create a PR: https://github.com/plbossart/sound/pull/9  to remove these check. There is no regression issue.  Do you  prefer this defensive code or no check ? Thanks!

plbossart replied:
@RanderWang your idea of not waking up ghost devices does not work with this code.

when a device is probe, we read the properties, and that's how we set wake-capable.
	/* device is probed so let's read the properties now */
	if (slave->ops && slave->ops->read_prop)
		slave->ops->read_prop(slave);

In other words, even ghost devices are seen as capable of generating wakes.

RanderWang replied:
@plbossart    Now I check slave->dev_num to skip ghost devices, please check https://github.com/plbossart/sound/pull/11. Thanks!

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -344,6 +344,21 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+	/*
+	 * When soundwire is in clock stop state, slave will wake up
+	 * master for some events like jack plugged and wakeen interrupt
+	 * will be triggered by audio pci device and routed to PME if pci
```
@plbossart this change seems to confusing to follow(probably because I am not very familiar with the way SDW works) but here's my concern. 
The above statement says that "wakeen interrupt  will be triggered by audio pci device". Is this really true? Isnt the wakeen interrupt triggered by the glue hardware that has received the jack detection event? Which in turn resumes the SOF PCI device? In which case, we should handle this event in the resume callback instead of when we boot the firmware? 
Maybe I am off track but I thought I'll ask anyway as this is quite different from how we handle the WAKEEN feature for HDA codecs.

ranj063 replied:
And oh the commit title prefix needs to be fixed as well

plbossart replied:
reworded everything and fixed title

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -345,17 +345,19 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+	 * to a WAKEEN interrupt, handled by the PCI device and routed
+	 * to PME if the PCI device is in D3. The resume function in
+	 * audio PCI driver will be invoked by ACPI for PME event and
+	 * initialize the device and process WAKEEN interrupt.
```
Ahh this makes a lot more sense. Then is the only reason to handle the event here is to wait until after ROM initialization?

plbossart replied:
yes, made it clear in the text

ranj063 commented on drivers/soundwire/intel.c:
```diff
@@ -1427,6 +1428,16 @@ static int intel_master_startup(struct sdw_master_device *md)
+		 * reference count
+		 */
+		pm_runtime_get_noresume(&md->dev);
+	}
```
@plbossart why not club this with the checks in line 1422 and not even enable runtime  PM?

plbossart replied:
added more explanations.

This is different, the quirk is specified by the parent while disabling pm_runtime is specified by the user for testing only. We want to keep the two separate.

ranj063 commented on drivers/soundwire/intel_init.c:
```diff
@@ -297,6 +307,16 @@ sdw_intel_startup_controller(struct sdw_intel_ctx *ctx)
+			 * power rails to the SoundWire IP are not
+			 * turned off.
+			 */
+			pm_runtime_get_noresume(link->dev);
```
@plbossart this will never happen in our HDA platforms right? Because we always default to BUS_RESET?
This change makes me very uncomfortable for some reason.

plbossart replied:
no, we will need to use quirks == 0 when we need to do acoustic event detection.

ranj063 commented on drivers/soundwire/bus.c:
```diff
@@ -1239,20 +1239,27 @@ static int sdw_handle_slave_alerts(struct sdw_slave *slave)
+	ret = pm_runtime_get_sync(&slave->dev);
+	if (ret < 0 && ret != -EACCES) {
+		dev_err(&slave->dev, "Failed to resume device: %d\n", ret);
+		pm_runtime_put_noidle(slave->bus->dev);
```
just wondering if the modify_slave_status() just above needs to be addressed if the device failed to resume?

plbossart replied:
no, the slave status is what we see on the bus, it's different from the status of the Linux device.

The English language does not have enough terms for 'device'

ranj063 commented on drivers/soundwire/bus.c:
```diff
@@ -1239,20 +1239,27 @@ static int sdw_handle_slave_alerts(struct sdw_slave *slave)
 
 	sdw_modify_slave_status(slave, SDW_SLAVE_ALERT);
 
+	ret = pm_runtime_get_sync(&slave->dev);
```
Also typos and grammar mistakes in the commit message. 
"There are two types of io error*s* when processing alert event.
The first one is: master receives a*n* alert event for jack event
and invokes implement_def function in slave to check jack status.
At this time codec is just suspended, so io registers can't be
accessed. Another one is: when w*o*ken *or waking* up from clock stop state
and bus needs a complete re-enumeration for synchronization issue
, slave register can't be accessed."

plbossart replied:
I did not modify this commit message since it's a controversial patch, with errors reported in https://github.com/thesofproject/linux/issues/1597

@RanderWang we really need to figure out if this is needed or not.

RanderWang replied:
@plbossart   I showed the details in https://github.com/thesofproject/linux/pull/1589. Thanks!

RanderWang replied:
> implement_def

Thanks! @ranj063  I fixed in https://github.com/thesofproject/linux/pull/1589

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+		val = sdw_bread_no_pm(bus, dev_num, SDW_SCP_STAT) &
+			SDW_SCP_STAT_CLK_STP_NF;
+		if (!val)
+			break;
```
Can we return 0 here?

plbossart replied:
you'd miss the dev_info log below

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+		ret = sdw_slave_clk_stop_prepare(slave,
+						 slave_mode, true);
+		if (ret < 0) {
+			dev_err(&slave->dev,
```
Should we do something when sdw_slave_clk_stop_callback() succeed, but sdw_slave_clk_stop_prepare() fail? And the case that sdw_slave_clk_stop_callback() fail for a slave, but succeed for previous slaves.

plbossart replied:
yes we could.

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+			continue;
+
+		/* Identify if Slave(s) are available on Bus */
+		is_slave = true;
```
Should we set ``is_slave = true;`` after 
```
if (slave->status != SDW_SLAVE_ATTACHED &&
		    slave->status != SDW_SLAVE_ALERT)
			continue;
```
?

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+
+		if (slave->status != SDW_SLAVE_ATTACHED &&
+		    slave->status != SDW_SLAVE_ALERT)
+			continue;
```
Should we move ``is_slave = true;`` here?

plbossart replied:
yes

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+
+	if (is_slave && !simple_clk_stop)
+		sdw_bus_wait_for_clk_prep_deprep(bus, SDW_BROADCAST_DEV_NUM);
+
```
Would it be better to test is_slave before list_for_each_entry(slave, &bus->slaves, node) {}?

plbossart replied:
yes

bardliao commented on drivers/soundwire/bus.c:
```diff
@@ -687,6 +734,298 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
+			continue;
+
+		slave_mode = slave->curr_clk_stop_mode;
+
```
Would it be better to test is_slave before list_for_each_entry(slave, &bus->slaves, node) {}? ie. Don't do list_for_each_entry(slave, &bus->slaves, node) {} if is_slave is false.

plbossart replied:
yes

plbossart commented:
all merged in various branches already
